### ISSUE TITLE
What kind of code we would like to be able to write

### ISSUE THREAD

**Issue by lassik** [pre-srfi/http-client-serial#2]
This is a complex topic; it may be best to start design by writing examples of clean code that does something practical.

**Comment by lassik** [comment]
To start with, I'd like these.

GET:

```
(http-get-string "https://example.com/")
(http-get-bytevector "https://example.com/foo.zip")
```

POST:

```
(http-post-string "https://example.com/hello" "Hello world!")
(http-post-bytevector "..." "...")
```

The POST ones may need to specify an explicit content-type in the header most of the time.

**Comment by lassik** [comment]
It should be possible to pass key-value pairs (from association lists, and maybe other collections) into the URL query part and/or the POST/PUT request body. But not all web services expect the request body as HTML-form-encoded key-value pairs, so we need the ability to send a raw bytevector/string as well.

**Comment by johnwcowan** [comment]
The query-part of an HTTP(s) URL is always name-value pairs.  The *body* can have any media type including "application/x-www-form-urlencoded", and perhaps there should be some way to convert an alist or other dictionary into this format.  But it has nothing to do with HTTP as such.

We need to support all the verbs.  At work I am currently rewriting a servlet that until now only accepted GET and POST to use, as appropriate, GET, POST, DELETE, and PATCH (it has no need for PUT but there are other contexts for it).  OPTIONS and HEAD are also standardized but there is less need for them.

As for fetching strings, there is no guarantee whether the result is textual, and even the media type doesn't tell you that unless it's one you recognize.  There is also the problem that we don't know in advance how much data the server will return.  That's why I designed my HttpClient interface to take an accumulator of bytevectors to send to the server and a generator of bytevectors received from the server.  It is then easy to append the bytevectors together and/or convert to strings according to the encoding; this can be done at a higher level than plain HTTP client operations.

A general principle here: rather than passing or receiving a zillion options, package just the lowest level and then easily use higher levels.

**Comment by lassik** [comment]
> The query-part of an HTTP(s) URL is always name-value pairs.

I recall that isn't always so, but a quick search doesn't turn up anything other than vague allusions.

> The _body_ can have any media type including "application/x-www-form-urlencoded", and perhaps there should be some way to convert an alist or other dictionary into this format. But it has nothing to do with HTTP as such.

I think the SRFI should provide a symmetrical way to pack an alist/dictionary into the body and into the query string. HTML forms can be packed into either; it's confusing to people who aren't HTTP experts if the two cases are different.

> We need to support all the verbs.

Agreed. Indeed, for the low-level operations the verb (aka method) should probably be orthogonal to every other part of the request object.

> As for fetching strings, there is no guarantee whether the result is textual, and even the media type doesn't tell you that unless it's one you recognize. There is also the problem that we don't know in advance how much data the server will return.

We should still provide the convenience; it's a godsend for scripting. There can be safeguards like a max limit on response body length, and a failed string decode can raise an exception or fill in invalid-character codepoints.

We can put the convenience in a separate SRFI if desired.

> That's why I designed my HttpClient interface to take an accumulator of bytevectors to send to the server and a generator of bytevectors received from the server. It is then easy to append the bytevectors together and/or convert to strings according to the encoding; this can be done at a higher level than plain HTTP client operations.

Generators and accumulators are nice but we should reconsider Feeley's points about ports and timeouts as well.

> A general principle here: rather than passing or receiving a zillion options, package just the lowest level and then easily use higher levels.

Unfortunately, the "easily" part is forgotten by lots of HTTP client libraries. That's why the Python `requests` library was such a hit.

**Comment by johnwcowan** [comment]
You're correct about query-parts, of course.  The htmlpreview.github.io service takes an URL directly after the question mark.  But I don't know what you mean by "symmetrical"; one is part of an URL creator, the other is a way of creating an entity body.  I agree that both should accept dictionaries.

I used `requests` as the initial model for HttpClient.

**Comment by lassik** [comment]
I mean that all of these should work intuitively (these examples use Common Lisp syntax for keyword args):

```
(make-http-url :query "foo")
(make-http-url :query '(("name1" . "value1") ("name2" . "value2")))

(make-http-request :body "foo")
(make-http-request :body '(("name1" . "value1") ("name2" . "value2")))
```

It might make sense to require some other arguments in addition to these ones, to say "please use HTML form encoding" or "please use variant foo of URL encoding" or something like that.

**Comment by lassik** [comment]
Here's some real Emacs Lisp code from Arthur's validate-html package (lightly edited for context):

```
(let* ((url-request-method "POST")
       (url-request-data (encode-coding-string (buffer-string) 'utf-8))
       (url-request-extra-headers
        `(("Content-Type" . "text/html; charset=utf-8"))))
  (with-temp-buffer
    (url-insert-file-contents
     "https://validator.w3.org/nu/?out=json&level=error")
    (json-read)))
```

**Comment by johnwcowan** [comment]
The first three look fine.  But the fourth: should it be sent www-url-form-encoded or JSON?  The first is right if you are talking to a conventional web server, the second is right if you are talking to an application server.  So either we can add a new option (which quickly becomes a nightmare) or we can write:

```
(make-http-request :url whatever 
                   :media-type "application/json"
                   :body (to-json alist))

(make-http-request :url whatever 
                   :media-type "application/x-www-form-urlencoded"
                   :body (to-www-form alist))
```

This looks redundant, but I don't want `make-http-request` to decipher its `:media-type` argument and use that to try to encode whatever `:body` is.  Instead, make `:body` accept a string or a bytevector, period, and leave other encodings to be converted using functions as shown here.  The strings might be bound to constants to make them less error-prone.

"Feeley's points about ports and timeouts":  pointer?

(It's annoying that SRFI 180 doesn't have simple `to-json` and `from-json` procedures.)

Thing to note:  about 95.5% of all websites are now UTF-8.

**Comment by lassik** [comment]
> This looks redundant, but I don't want `make-http-request` to decipher its `:media-type` argument and use that to try to encode whatever `:body` is. Instead, make `:body` accept a string or a bytevector, period, and leave other encodings to be converted using functions as shown here. The strings might be bound to constants to make them less error-prone.

That's a reasonable approach.

Another alternative is to have procedures that modify a request object like this:

```
(http-request-set-query (make-http-request) '(("key1" . "value1") ("key2" . "value2")))

(http-request-set-body-form (make-http-request) '(("key1" . "value1") ("key2" . "value2")))
(http-request-set-body-json (make-http-request) '(("key1" . "value1") ("key2" . "value2")))
```

They could mutate the given request, or return a copy of that request with the requested changes.

I'm not sure which kind of API I prefer yet. Yours is also nice.

> "Feeley's points about ports and timeouts": pointer?

We had a big discussion on the SRFI lists sometime early this year (or late last year) about the virtues of generators and accumulators vs ports. I can't remember the context; most likely it was about how to return directory listings. Feeley opined that ports are a superior abstraction because they tie into the Scheme implementaiton's I/O framework so you can specify timeouts and other options for read and write operations.

> (It's annoying that SRFI 180 doesn't have simple `to-json` and `from-json` procedures.)

Library design is just too hard. All other languages have the same problem. Common Lisp programmers continually consider updating the ANSI standard but can never come to an agreement on how to do it. We should figure out some way to do those yearly editions at some point. Scheme 2020, Scheme 2021, etc. Probably using a less controversial name.

> Thing to note: about 95.5% of all websites are now UTF-8.

Which is great. I just wish more software was updated with sane defaults to reflect this fact. I continually run into misconfiguration problems with UTF-8 text.

**Comment by johnwcowan** [comment]
> Another alternative is to have procedures that modify a request object like this:

That's the Builder pattern, which I consider to be an anti-pattern for Lisp-style languages even if the object is immutable, because it requires extending the API surface when new components are needed.  What we want in order to construct hairy objects like HTTP request is ... dum de dum dum ... keywords, in the form of function-call keywords, plists, or alists.


> We had a big discussion on the SRFI lists sometime early this year (or late last year) about the virtues of generators and accumulators vs ports. I can't remember the context; most likely it was about how to return directory listings. Feeley opined that ports are a superior abstraction because they tie into the Scheme implementaiton's I/O framework so you can specify timeouts and other options for read and write operations.

Ah yes, I remember.  If we were willing to tie this SRFI to  SRFI 181 (custom ports) we could easily do that.  I'm not sure whether that's the Right Thing or not.

> Library design is just too hard. All other languages have the same problem. 

Eh, adding a supplementary SRFI to be incorporated in the same library would be easy.

**Comment by amirouche** [comment]
On the subject of generator vs. accumulators:

> Feeley opined that ports are a superior abstraction because they tie into the Scheme implementaiton's I/O framework so you can specify timeouts and other options for read and write operations.

Good argument. Also, I tested ports vs. generators in the context of reading a local file using Chez Scheme, ports are faster (ports are faster than mmap'ing the whole file and they stream it).

**Comment by amirouche** [comment]
I like the API of chicken [http-client](http://wiki.call-cc.org/eggref/5/http-client#simple-get-request). Here is an example:

```scheme
(call-with-output-file
  "/tmp/target.png"
  (lambda (out)
    (call-with-input-request
      "http://example.com/file.png"
      #f                                                                                                                                  
      (lambda (in) (copy-port in out)))))
```

Though, if we just want to support the simple case where the client sends a small body, we can simplify into:

```scheme
(call-http method mime-type bytevector proc)
```

Where `PROC` takes as arguments:

- status code
- headers
- a binary port

A simple implementation would send the header "close" header. Also, the port would produce the body without chunks extensions and trailing headers.
